src/validationinterface.cpp

```diff
@@ -214,15 +205,38 @@ void ThreadNotifyWallets(CBlockIndex *pindexLastTip)
                         OrchardMerkleFrontier::empty_root(), oldFrontiers.orchard));
                 }
 
+                // Fetch recently-conflicted transactions. These will include any
+                // block that has been connected since the last cycle, but we only
+                // notify for the conflicts created by the current active chain.
+                auto recentlyConflicted = TakeRecentlyConflicted(pindex);
+
                 blockStack.emplace_back(
                     pindex,
                     oldFrontiers,
-                    recentlyConflicted.first[pindex]);
+                    recentlyConflicted.first);
+
+                chainNotifiedSequence = recentlyConflicted.second;
 
                 pindex = pindex->pprev;
             }
 
-            recentlyAdded = mempool.DrainRecentlyAdded();
+            // This conditional can be true in the case that in the interval
+            // since the last second-boundary, two reorgs occurred: one that
+            // shifted over to a different chain history, and then a second
+            // that returned the chain to the original pre-reorg tip.  This
+            // should never occur unless a caller has manually used
+            // `invalidateblock` to force the second reorg or we have a long
+            // persistent set of dueling chains. In such a case, wallets may
+            // not be fully notified of conflicted transactions, but they will
+            // still have a correct view of the current main chain, and they
+            // will still be notified properly of the current state of
+            // transactions in the mempool.
+            if (originalTipAtFork) {
+                chainNotifiedSequence = GetChainConnectedSequence();
```

Comment:
I don't understand what goes wrong if we don't do this. It seems like `originalTipAtFork` will only catch a subset of possible multiple-reorg situations, where the chain returns *exactly* to the original tip. If that would cause a problem, then why wouldn't (for example) returning to the original tip plus one extra block also cause a problem?

Reply:
Oh, is it that the tests specifically depend on that situation?

Reply:
Yep. RPC tests using "clean" chains only start with the genesis block, so this boolean is true (everything is synced to genesis), but ConnectTip was being called for the genesis block in some pathway causing the sequence number to be incremented. Previously this was handled because we always called `Drain...` even if the tips were the same, so we fetched the latest sequence number.

Reply:
The error that this addresses is one that was caught by the RPC tests, which is that for the genesis block `ConnectTip` *is* called, but `ThreadNotifyWallets` doesn't learn of it. In the case of a multiple-reorg situation where anything else is returned (`pindex != pindexFork`, and therefore `originalTipAtFork == false`) the loop above will set `chainNotifiedSequence` correctly.